service-mesh-probe/service-mesh.proto

```diff
@@ -59,6 +59,13 @@ message ServiceMeshMetric {
 
     // The tcp detail of the mesh metrics if it's a TCP metrics, otherwise should be empty.
     TCPInfo tcp = 15;
+
+    // Internal request(response) latency is the time cost inside the sidecar proxy
+    // after it received the request(response) from the downstream(upstream) and
+    // before it sents the request(response) to upstream(downstream).
+    // The time typically is costed by the sidecar proxy filters.
+    int64 internalRequestLatencyNanos = 16;
+    int64 internalResponseLatencyNanos = 17;
```

Comment:
```suggestion
    // Internal request(response) latency is the time cost inside the sidecar proxy
    // The time typically is costed by the sidecar proxy filters.
    // Latency = timestamp of first request package received - timestamp of first package sent to upstream.
    int64 internalRequestLatencyNanos = 16;
    // Latency = timestamp of first response package received - timestamp of first package sent to downstream.
    int64 internalResponseLatencyNanos = 17;
```

Reply:
@kezhenxu94 I think these would be a more clear definition.

Reply:
@wu-sheng I added some more details with ascii graph that clearly state how we calculate the 2 metrics

Reply:
<img width="939" alt="image" src="https://user-images.githubusercontent.com/5441976/166928264-c0d57707-da95-447c-927c-3e93075ed71e.png">

This is what I see from the review page. Are these variables placed in the right place?
`time_to_first_upstream_tx_byte` should be placed between proxy to upstream? I think maybe ` ascii graph` would show correct, but it seems strange in the comments.

Reply:
> `time_to_first_upstream_tx_byte` should be placed between proxy to upstream?

`time_to_first_upstream_tx_byte` is the interval between the first downstream byte received and the first upstream byte sent, it doesn't include the time used to send the first byte to upstream

Reply:
OK, I just rechecked the doc

> time_to_first_upstream_tx_byte
([Duration](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#duration)) Interval between the first downstream byte received and the first upstream byte sent. There may by considerable delta between time_to_last_rx_byte and this value due to filters. Additionally, the same caveats apply as documented in time_to_last_downstream_tx_byte about not accounting for kernel socket buffer time, etc.

As it is already duration, it is wrong `internalRequestLatencyNanos = time_to_first_upstream_tx_byte - time_to_last_rx_byte`, then **internalRequestLatencyNanos = time_to_first_upstream_tx_byte**.

Reply:
I think we could mention `internalRequestLatencyNanos = time_to_first_upstream_tx_byte.` and add link to https://www.envoyproxy.io/docs/envoy/latest/api-v3/data/accesslog/v3/accesslog.proto

Reply:
No, even it's already duration, it's wrong `internalRequestLatencyNanos = time_to_first_upstream_tx_byte`, please check my description

```
    // | client  | ------- request --------> |                proxy                | --------- request -------> | upstream |
    //           |    time_to_last_rx_byte   |
    //           |                time_to_first_upstream_tx_byte                   |
    //                                       |     internalRequestLatencyNanos     |
```

all `time_to_xxxxxx` are relative to `the first downstream byte received`, so you'll see they are cumulative.

```
| `the first downstream byte received`
|                      time_to_xxxxxxxxxx |
|                      time_to_xxxxxxxxxx |       time_to_yyyyyyyyyyyyy      |
|                      time_to_xxxxxxxxxx |       time_to_yyyyyyyyyyyyy      |     time_to_zzzzzzzzzzzz     |
```

Reply:
From my understanding, `time_to_first_upstream_tx_byte` is the whole latency. `time_to_last_rx_byte` should be included, because the packages receiving is a part of sidecar/envoy costs too.